src/MongoDB.Driver.Core/Core/Compression/Zstandard/ZstandardStream.cs

```diff
@@ -0,0 +1,369 @@
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+
+            if (!_disposed)
+            {
+                switch (_compressionMode)
+                {
+                    case CompressionMode.Compress:
+                        if (!_flushed)
+                        {
+                            try { Flush(); } catch { }
+                        }
+                        _nativeWrapper.Dispose();
+                        _arrayPool.Return(_streamWriteHelper.CompressedBufferInfo.Bytes);
+                        break;
+                    case CompressionMode.Decompress:
+                        _nativeWrapper.Dispose();
+                        _arrayPool.Return(_streamReadHelper.CompressedBufferInfo.Bytes);
+                        break;
+                }
+
+                _disposed = true;
+            }
+        }
+
+        public override void Flush()
+        {
+            foreach (var outputBufferPosition in _nativeWrapper.StepwiseFlush(_streamWriteHelper.CompressedBufferInfo))
+            {
+                _streamWriteHelper.WriteBufferToCompressedStream(count: outputBufferPosition);
+            }
+            _compressedStream.Flush();
+            _flushed = true;
+        }
+
+        public override int Read(byte[] outputBytes, int outputOffset, int count) // Decompress
+        {
+            if (!CanRead) throw new InvalidDataException("Read is not accessible.");
+
+            var uncompressedOutputBuffer = new BufferInfo(outputBytes, outputOffset);
+
+            using (var operationContext = _nativeWrapper.InitializeOperationContext(
+                _streamReadHelper.CompressedBufferInfo,
+                uncompressedOutputBuffer)) // operation result
+            {
+                var length = 0; // the result size
+                var remainingCount = count;
+
+                while (remainingCount > 0)
+                {
+                    var currentDataPosition = _streamReadHelper.ReadCompressedStreamToBufferIfAvailable(
+                        recommendedCompressedSize: _nativeWrapper.RecommendedInputSize,
+                        out int remainingCompressedBufferSize);
+
+                    // decompress input to output
+                    _nativeWrapper.Decompress(
+                        operationContext,
+                        compressedOffset: currentDataPosition,
+                        inputCompressedSize: remainingCompressedBufferSize,
+                        inputUncompressedSize: remainingCount,
+                        out int compressedBytesProcessed,
+                        out int uncompressedBytesProcessed);
+
+                    if (!_streamReadHelper.TryPrepareDataForNextAttempt(
+                        compressedBytesProcessed,
+                        uncompressedBytesProcessed))
+                    {
+                        break;
+                    }
+
+                    length += uncompressedBytesProcessed;
```

Comment:
![image](https://user-images.githubusercontent.com/44870738/77087414-27551900-6a14-11ea-9270-7a172b300483.png)

Reply:
Would `totalUncompressedBytesProcessed` capture the meaning of this variable?

Reply:
i don't mind. Done